### DISCUSSION TITLE
Crop images instead of distorting size ratios

### DISCUSSION THREAD

**Discussion by nicholaschiang** [vercel/next.js#18871] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Right now, when specifying an image `width` and `height` that don't match the actual image source leads to [horrible image distortion](https://develop.tutorbook.app/default/search/ztv6sQzQv8Ov9u1UlCy4Y9VQ9F13?query=&orgs=%255B%257B%2522value%2522%253A%2522gunn%2522%252C%2522label%2522%253A%2522Gunn%2520High%2520School%2522%257D%255D&tags=%255B%255D&page=0&hitsPerPage=20&aspect=tutoring&langs=%255B%255D&subjects=%255B%255D&availability=%255B%255D&checks=%255B%255D&parents=%255B%255D&visible=true) (because Next.js just updates styling on a wrapper `div` and doesn't actually crop the image at runtime):

![image](https://user-images.githubusercontent.com/20798889/98309281-ec8dfa00-1f7e-11eb-9b57-b4fae638ee0a.png)

The [actual image source](https://develop.tutorbook.app/_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fproduction-tutorbook.appspot.com%2Fo%2Ftemp%252F84c32320-585f-4cc4-b6d2-5c7555bc4d3c.jpg%3Falt%3Dmedia%26token%3D4c6b69ae-63a9-4ce5-b97c-58a37ac279e0&w=260&q=75) doesn't have the requested aspect ratio (instead, Vercel just resizes the image `width` at runtime and leaves the original ratio intact):

![image](https://user-images.githubusercontent.com/20798889/98309378-1c3d0200-1f7f-11eb-932e-3637dd125f0f.png)

[Here is my code](https://github.com/tutorbookapp/tutorbook/blob/develop/components/avatar/index.tsx#L46) that uses the new `<Image />` component for those performance optimizations (and results in the distorted image sizing depicted above):

```tsx
export default function Avatar({
  src = '',
  loading = false,
  size = 500,
  verified,
}: AvatarProps): JSX.Element {
  const { t } = useTranslation();

  return (
    <div className={cn(styles.wrapper, { [styles.loading]: loading })}>
      {verified && (
        <Tooltip
          content={<div className={styles.verifiedHover}>{verified}</div>}
          align='right'
          showArrow
        >
          <div className={styles.verifiedText}>{t('common:verified')}</div>
        </Tooltip>
      )}
      {!loading && !!src && (
        <Image data-cy='avatar' height={size} width={size} src={src} alt='' />
      )}
      {!src && !loading && (
        <div className={styles.noPhotoWrapper}>
          <div className={styles.noPhoto}>{t('common:no-photo')}</div>
        </div>
      )}
    </div>
  );
}
```

## Describe the solution you'd like

Vercel should use something like [`smartcrop.js`](https://github.com/jwagner/smartcrop-sharp) to dynamically crop those images to fit the specified `width` and `height` (in addition to the existing resizing and WEBP format conversions).

It'd also be really nice to have [optional face recognition](https://29a.ch/sandbox/2014/smartcrop/examples/testbed.html) as well using something like [`opencv`](https://github.com/peterbraden/node-opencv) or [`tracking.js`](https://trackingjs.com/).

[Here's an example](https://github.com/jwagner/smartcrop-cli/blob/master/smartcrop-cli.js) of smart facial recognition image cropping on a Node.js runtime. Note that you'd have to have `imagemagick` or [`sharp`](https://github.com/lovell/sharp) installed for some of this functionality (though I'm assuming Vercel already has a similar setup for the existing runtime image cropping).

## Describe alternatives you've considered

Right now, I'm doing this (cropping images to a square 500x500 ratio) on the server-side and then supplying Vercel with the link to the cropped image. It'd be nice to have these features built-in at some point.

**Comment by jose-verissimo** [comment]
The image resize seems very pointless if the images are distorted, this is happening at my end, wanting to transform an image from landscape to a square ratio but it's being distorted. 

I'll crop it manually for now, but it would be a nice feature to add when you guys have time.

A good thing to add as well would be a a focal point to where the transform happens, so like:
- Top-Left
- Top-Center
- Top-Right
- Center-Left
- Center-Center
- Center-Right
- Bottom-Left
- Bottom-Center
- Bottom-Right

The mode property could be set to either 'crop', 'fit', or 'stretch' to the desired width/height.
If the mode is set to 'crop', you can pass a position property, set to either 'top-left', 'top-center', 'top-right', 'center-left', 'center-center', 'center-right', 'bottom-left', 'bottom-center', or 'bottom-right'.

Hope that makes sense, in 2021 I should have a lot more time so I'll look into contributing if this hasn't been implemented yet.

**Comment by nicholaschiang** [comment]
Related to tutorbookapp/tutorbook#139 (for anyone who wants to check out the original "workaround" serverless solution).

**Comment by jacksonzumdish** [comment]
PHP Glide has the features we need, namely cropping, percentage based focal crop, etc. I know it's kinda antithetical to next.js but it would be a great place to start in terms of feature parity...

https://glide.thephpleague.com

**Comment by siakar** [comment]
Hi, how did you do it(cropping images to a square 500x500 ratio) on the server-side?
